Add support and tests for reference only builds

marknfawaz opened the pull request:
## Related issue

Closes: #71

Eruanion reviewed (changes requested):
There are build warnings in src/Analysis/Codelyzer.Analysis.Model/InvocationExpression.cs:
'InvocationExpression.Parameters' is obsolete: 'Use Argument property instead. This property will be removed in a later version.'

src/Analysis/Codelyzer.Analysis.CSharp/SemanticHelper.cs : Is also getting unused variable warning in case we wanted to add those errors to logging.

marknfawaz commented:
> There are build warnings in src/Analysis/Codelyzer.Analysis.Model/InvocationExpression.cs:
> 'InvocationExpression.Parameters' is obsolete: 'Use Argument property instead. This property will be removed in a later version.'
> 
> src/Analysis/Codelyzer.Analysis.CSharp/SemanticHelper.cs : Is also getting unused variable warning in case we wanted to add those errors to logging.

We added these warnings to warn users using "Parameters" that this will be removed in the future and replaced with Arguments because it contains more data.

nshiramshetti commented on src/Analysis/Codelyzer.Analysis.Build/ProjectBuildHandler.cs:
```diff
@@ -444,139 +450,32 @@ public ProjectBuildResult SyntaxOnlyBuild()
         }
         private void GetTargetFrameworks(ProjectBuildResult result, Buildalyzer.IAnalyzerResult analyzerResult)
```
nit: Set method

nshiramshetti commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilder.cs:
```diff
@@ -56,5 +56,28 @@ public async Task<List<ProjectBuildResult>> Build()
             return ProjectResults;
         }
 
+        public async Task<List<ProjectBuildResult>> GenerateNoBuildAnalysis(Dictionary<string, List<string>> oldReferences, Dictionary<string, List<string>> references)
```
At a high level, how do we track the issues from build using references?

marknfawaz replied:
Using the same method as a normal compilation. I will add the logic for this.

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/ProjectBuildHandler.cs:
```diff
@@ -134,7 +135,7 @@ private async Task SetCompilation()
                     Console.WriteLine(e);
```
Yes. Removed.

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/ExternalReferenceLoader.cs:
```diff
@@ -0,0 +1,215 @@
+                var packageFolderDir = Path.Combine(packageDirSplit[0], Constants.PackagesFolder, packageFolderName);
+                var nupkg = Directory.EnumerateFiles(packageFolderDir, Constants.NupkgFileExtension, SearchOption.AllDirectories).ToList();
+                nupkg.ForEach(file => {
+                    var reader = new PackageArchiveReader(file);
```
We are handling it on the individual package level (the function calling this). Any exception in this will cause a single nuget parsing to fail and continue with the remaining ones. I removed the foreach here because a nuget will have one nupkg file only (the foreach was misleading).

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/ExternalReferenceLoader.cs:
```diff
@@ -0,0 +1,215 @@
+                    }
+                    catch (Exception ex)
+                    {
+                        Logger.LogError(ex, "Error while resolving reference {0}", externalReferenceMetaData);
```
We are doing this indirectly. The logger being used here is passed to Codelyzer from the client application.

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/ExternalReferenceLoader.cs:
```diff
@@ -0,0 +1,215 @@
+            {
+                _packageReferences?.ToList().ForEach(packageReference => {
+                    var reference = new ExternalReference() {
+                        Identity = packageReference.Key,
```
Not sure I get this comment. We already are checking. In fact, we have an additional null check that's not needed here.

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/ExternalReferenceLoader.cs:
```diff
@@ -0,0 +1,215 @@
+                projectReferenceNames = projectReferences != null ? projectReferences.Select(p => p.Name).ToHashSet<string>() : null;
+
+                _externalReferences.ProjectReferences.AddRange(projectReferences.Select(p => new ExternalReference()
+                {
```
Both of these cannot be null: _externalReferences is initialized in the class constructor and ProjectReferences is initialized in its class constructor.

nshiramshetti reviewed (commented):
Could you please resolve all the warnings?

marknfawaz commented:
> Could you please resolve all the warnings?

We will remove the warnings in a later version. These are meant to warn users to stop using Parameters and start using Arguments (breaking change).

nshiramshetti commented:
Sounds good

On Tue, May 25, 2021 at 4:31 PM Mark Fawaz ***@***.***> wrote:

> Could you please resolve all the warnings?
>
> We will remove the warnings in a later version. These are meant to warn
> users to stop using Parameters and start using Arguments (breaking change).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/aws/codelyzer/pull/72#issuecomment-848337197>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQC4ONHRG5EULOBMZBTRXTDTPQXNHANCNFSM44XNHKGQ>
> .
>
